From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: quh4gko8 <88831734+quh4gko8@users.noreply.github.com>
Date: Wed, 20 May 2026 14:40:16 +0000
Subject: [PATCH] Add permission checks for Motion Sensors site settings

Fixup with in future releases:

fixup! Initial support for toggling Sensors site settings per-site
---
 .../site_settings/SensorsSettingsHooks.java   | 86 +++++++++++++++++++
 .../SiteSettingsHooksRegistry.java            |  2 +
 2 files changed, 88 insertions(+)

diff --git a/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SensorsSettingsHooks.java b/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SensorsSettingsHooks.java
index 817c2976b1e26..5ebbc140a26c2 100644
--- a/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SensorsSettingsHooks.java
+++ b/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SensorsSettingsHooks.java
@@ -1,5 +1,13 @@
 package org.chromium.components.browser_ui.site_settings;
 
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.os.Process;
+
+import androidx.preference.Preference;
+
+import org.chromium.base.ApiCompatibilityUtils;
+import org.chromium.base.ContextUtils;
 import org.chromium.build.annotations.NonNull;
 import org.chromium.build.annotations.NullMarked;
 import org.chromium.build.annotations.Nullable;
@@ -7,9 +15,69 @@ import org.chromium.components.content_settings.ContentSetting;
 import org.chromium.components.content_settings.ContentSettingsType;
 import org.chromium.content_public.browser.BrowserContextHandle;
 
+import app.vanadium.config.VanadiumConfParser;
+import app.vanadium.ext.CustomOSPermissions;
+
+import static app.vanadium.config.proto.VanadiumConfigProto.Flag.FlagType.DISABLED_FEATURE;
+
 @NullMarked
 public final class SensorsSettingsHooks {
 
+    private static boolean isPermissionExistingAndRevoked() {
+        String featureKey = "CheckOSPermissionForSensorsSiteSettingsUI";
+        if (VanadiumConfParser.getFlagsFor(DISABLED_FEATURE).containsKey(featureKey)) {
+            return false;
+        }
+
+        String permission = CustomOSPermissions.getSensorsPermissionName();
+        if (permission == null) {
+            return false;
+        }
+
+        return ApiCompatibilityUtils.checkPermission(
+                ContextUtils.getApplicationContext(), permission,
+                Process.myPid(), Process.myUid()) != PackageManager.PERMISSION_GRANTED;
+    }
+
+    static final class CategoryHooksImpl extends SiteSettingsCategoryHooks {
+
+        @Override
+        @NonNull
+        protected SiteSettingsCategory getSiteSettingsCategory(@NonNull BrowserContextHandle bch) {
+            return new SensorsSiteSettingsCategory(bch);
+        }
+    }
+
+    static final class SensorsSiteSettingsCategory extends SiteSettingsCategory {
+
+        SensorsSiteSettingsCategory(@NonNull BrowserContextHandle bch) {
+            super(bch, Type.SENSORS, "");
+        }
+
+        @Override
+        protected boolean enabledForChrome(Context context) {
+            return !isPermissionExistingAndRevoked();
+        }
+
+        @Override
+        protected boolean shouldDisableToggle() {
+            return isPermissionExistingAndRevoked();
+        }
+    }
+    static final class SiteSettingsScreenHooksImpl extends SiteSettingsScreenHooks {
+
+        @Override
+        protected void updateSiteSettingsPreference(
+                @NonNull SiteSettings siteSettingsFragment,
+                @NonNull SiteSettingsDelegate siteSettingsDelegate,
+                @NonNull Preference existingPreference) {
+            if (isPermissionExistingAndRevoked()) {
+                existingPreference.setSummary(ContentSettingsResources.getCategorySummary(
+                        ContentSetting.BLOCK, /* isOneTime= */ false));
+            }
+        }
+    }
+
     static final class WebsitePermissionsFetcherHooksImpl extends WebsitePermissionsFetcherHooks {
 
         @Override
@@ -35,6 +103,11 @@ public final class SensorsSettingsHooks {
             return isCategoryEnabled ? R.string.website_settings_add_site_description_motion_sensors_block
                                      : R.string.website_settings_add_site_description_motion_sensors_allow;
         }
+
+        @Override
+        protected boolean shouldAlwaysShowOsWarning() {
+            return isPermissionExistingAndRevoked();
+        }
     }
 
     static final class SingleWebsiteSettingsHooksImpl extends SingleWebsiteSettingsHooks {
@@ -44,6 +117,9 @@ public final class SensorsSettingsHooks {
                 @NonNull BrowserContextHandle browserContextHandle,
                 @Nullable PermissionSetting setting,
                 @ContentSettingsType.EnumType int contentType) {
+            if (isPermissionExistingAndRevoked()) {
+                return new PermissionSetting(null, ContentSetting.BLOCK, false);
+            }
             if (setting == null) {
                 boolean isEnabled = WebsitePreferenceBridge.isContentSettingEnabled(browserContextHandle, contentType);
                 int value = isEnabled ? ContentSetting.ALLOW : ContentSetting.BLOCK;
@@ -51,5 +127,15 @@ public final class SensorsSettingsHooks {
             }
             return setting;
         }
+
+        @Override
+        protected void setupOrModifyPreference(
+                @NonNull SingleWebsiteSettings siteSettingsFragment,
+                @NonNull BrowserContextHandle browserContextHandle,
+                @NonNull Preference preference) {
+            if (isPermissionExistingAndRevoked()) {
+                preference.setEnabled(false);
+            }
+        }
     }
 }
diff --git a/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SiteSettingsHooksRegistry.java b/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SiteSettingsHooksRegistry.java
index 91bcb464100f6..5e6b057f797f7 100644
--- a/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SiteSettingsHooksRegistry.java
+++ b/components/browser_ui/site_settings/android/java/src/org/chromium/components/browser_ui/site_settings/SiteSettingsHooksRegistry.java
@@ -35,6 +35,7 @@ public final class SiteSettingsHooksRegistry {
             @SiteSettingsCategory.Type int categoryType) {
         return switch (categoryType) {
             case SiteSettingsCategory.Type.JAVASCRIPT_JIT -> new JavascriptJitSettingsHooks.CategoryHooksImpl();
+            case SiteSettingsCategory.Type.SENSORS -> new SensorsSettingsHooks.CategoryHooksImpl();
             default -> SiteSettingsCategoryHooks.getDefaultInstance();
         };
     }
@@ -53,6 +54,7 @@ public final class SiteSettingsHooksRegistry {
             @SiteSettingsCategory.Type int categoryType) {
         return switch (categoryType) {
             case SiteSettingsCategory.Type.JAVASCRIPT_JIT -> new JavascriptJitSettingsHooks.SiteSettingsScreenHooksImpl();
+            case SiteSettingsCategory.Type.SENSORS -> new SensorsSettingsHooks.SiteSettingsScreenHooksImpl();
             default -> SiteSettingsScreenHooks.getDefaultInstance();
         };
     }
